torchtune/modules/transformer.py

```diff
@@ -502,29 +502,40 @@ def chunked_output(self, last_hidden_state: torch.Tensor) -> List[torch.Tensor]:
 
     def _validate_inputs(
         self,
-        seq_len: int,
+        tokens: Optional[torch.Tensor],
         mask: Optional[torch.Tensor] = None,
         encoder_input: Optional[torch.Tensor] = None,
         encoder_mask: Optional[torch.Tensor] = None,
         input_pos: Optional[torch.Tensor] = None,
+        input_embeds: Optional[torch.Tensor] = None,
     ):
         """
         Validates inputs for ``forward``.
         Args:
-            seq_len (int): Input tensor sequence length.
+            tokens (Optional[torch.Tensor]): input tensor with shape ``[b x s]``
             mask (Optional[torch.Tensor]): Attention mask used for inference and for sequence packing.
             encoder_input (Optional[torch.Tensor]): Encoder input for cross-attention.
             encoder_mask (Optional[torch.Tensor]): Encoder attention mask for cross-embedding attention.
             input_pos (Optional[torch.Tensor]): Input tensor position IDs.
+            input_embeds (Optional[torch.Tensor]): Input tensor embeddings (if short-circuiting token embeddings).
 
         Raises:
             ValueError:
+                If neither tokens nor input_embeds are passed **or**
                 If seq_len of x is bigger than max_seq_len, **or**
                 if the model has caches which have been setup with self-attention layers and ``mask`` is not provided, **or**
                 if the model has caches which have been setup with encoder layers and ``encoder_mask`` is not provided, **or**
                 if the model has caches which have been setup ``input_pos`` is not provided.
         """
 
+        if tokens is None and input_embeds is None:
```

Comment:
Should also check if both are not none - only one should be passed in

Reply:
Lol @joecummings too fast on the trigger finger merged before I even saw this. Can do as an easy follow-up